walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -55,4 +53,42 @@
     }
     return validTimes;
   }
+
+  public Collection<TimeRange> findOverlapTimes(Collection<TimeRange> mandatoryTimes, Collection<TimeRange> optionalTimes, long duration) {
+    Collection<TimeRange> validTimes = new ArrayList<>();
+    for(TimeRange optional : optionalTimes) {
+      for(TimeRange mandatory : mandatoryTimes) {
+        if(mandatory.overlaps(optional)) {
+          int start = Math.max(optional.start(), mandatory.start());
+          int end = Math.min(optional.end(), mandatory.end());
+          if(end == TimeRange.END_OF_DAY) {
+            end += 1;
+          }
+          TimeRange time = TimeRange.fromStartEnd(start, end, false);
+          if(time.duration() >= duration) {
+            validTimes.add(time);
+          }
+        }
+      }
+    }
+    return validTimes;
+  }
+
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> optional_attendees = request.getOptionalAttendees();
+    long duration = request.getDuration();
+    Collection<TimeRange> mandatoryTimes = findMeetingTimes(events, attendees, duration); //[]830-900
+    Collection<TimeRange> optionalTimes = findMeetingTimes(events, optional_attendees, duration); //[all day]
+    Collection<TimeRange> overlapTimes = findOverlapTimes(optionalTimes, mandatoryTimes, duration); //[]
+
+    if(overlapTimes.isEmpty()) {
+      if(attendees.isEmpty()) {
+        return optionalTimes;
+      }
+      return mandatoryTimes;
+    }
+    return overlapTimes;
+  }
```

Comment:
I personally find single level if-statement with logic operators easier to understand than nested if-blocks. I couldn't find any specification about it in the style guide though so up to you!

Reply:
Done. Also, I've seen it in a [TToT](https://drive.google.com/file/d/1iNgKtlzK6FnIUnSa87gfayfzrj56L3eb/view).